fri/src/verifier.rs

```diff
@@ -0,0 +1,282 @@
+    query_indices: Vec<u64>,
+    query_results: Vec<F>,
+    expected_last_layer: Option<Vec<F>>,
+    eval_points_test: Option<Vec<F>>,
+    last_layer_coefficients_test: Option<Vec<F>>,
+}
+
+#[allow(dead_code)]
+impl<F: FftField + PrimeField, P: Prng + Clone + 'static, W: Digest + Clone + 'static>
+    FriVerifier<F, P, W>
+{
+    pub fn new(
+        channel: FSVerifierChannel<F, P, W>,
+        params: FriParameters<F, Radix2EvaluationDomain<F>>,
+        commitment_hashes: CommitmentHashes,
+        first_layer_callback: FirstLayerQueriesCallback<F>,
+        eval_points_test: Option<Vec<F>>,
+        last_layer_coefficients_test: Option<Vec<F>>,
+    ) -> Self {
+        let n_layers = params.fri_step_list.len();
+        Self {
+            channel,
+            params,
+            commitment_hashes,
+            first_layer_callback,
+            n_layers,
+            first_eval_point: None,
+            eval_points: vec![F::zero(); n_layers - 1],
+            table_verifiers: vec![],
+            query_indices: vec![],
+            query_results: vec![],
+            expected_last_layer: None,
+            eval_points_test,
+            last_layer_coefficients_test,
+        }
+    }
+
+    pub fn verify_fri(&mut self) -> Result<(), Box<dyn Error>> {
+        self.commitment_phase()?;
+
+        // // query phase
+        //self.query_indices = self.choose_query_indices();
+        // //self.channel.begin_query_phase();
+
+        // // decommitment phase
+        // // TODO : annotation
+
+        // self.verify_first_layer();
+        // self.verify_inner_layers();
+        // self.verify_last_layer();
+        Ok(())
+    }
+
+    pub fn commitment_phase(&mut self) -> Result<(), Box<dyn Error>> {
+        let mut basis_index = 0;
+        for i in 0..self.n_layers {
+            let cur_fri_step = self.params.fri_step_list[i];
+            basis_index += cur_fri_step;
+
+            if i == 0 {
+                if self.params.fri_step_list[0] != 0 {
+                    self.first_eval_point = Some(self.channel.draw_felem());
+                }
+            } else {
+                self.eval_points[i - 1] = self.channel.draw_felem();
+            }
+
+            if i < self.n_layers - 1 {
+                let coset_size = 1 << self.params.fri_step_list[i + 1];
+                let n_rows = self.params.fft_domains[basis_index].size() / coset_size;
+                let n_columns = coset_size;
+                let size_of_row = ((F::MODULUS_BIT_SIZE.div_ceil(8) * 8) as usize) * n_columns;
+
+                let commitment_scheme = make_commitment_scheme_verifier(
+                    size_of_row,
+                    n_rows,
+                    0,
+                    self.commitment_hashes.clone(),
+                    n_columns,
+                );
+
+                let mut table_verifier = TableVerifier::new(n_columns, commitment_scheme);
+                let _ = table_verifier.read_commitment(&mut self.channel);
```

Comment:
Maybe `make_commitment_scheme_verifier` should take the commitment: is there any scenario where it makes sense to create a verifier without knowing which commitment it verifies against?

Reply:
It's confusing because TableVerifier uses 2-step initialization. Agreed.

Reply:
Maybe implementation below would be better?
(nit: is it an anti-pattern returning Resut<Self, Error> in ctor?)

```rust

    pub fn new(
        n_columns: usize,
        commitment_scheme: Box<dyn CommitmentSchemeVerifier<F, P, W>>,
        channel: &mut FSVerifierChannel<F, P, W>,
    ) -> Result<Self, anyhow::Error> {
        let mut verifier = Self {
            n_columns,
            commitment_scheme,
        };
        verifier.commitment_scheme.read_commitment(channel)?;
        Ok(verifier)
    }
```

from 

```rust
    /// Create new TableVerifier.
    pub fn new(
        n_columns: usize,
        commitment_scheme: Box<dyn CommitmentSchemeVerifier<F, P, W>>,
    ) -> Self {
        Self {
            n_columns,
            commitment_scheme,
        }
    }

    /// Reads the initial commitment into the scheme (e.g., Merkle root).
    pub fn read_commitment(
        &mut self,
        channel: &mut FSVerifierChannel<F, P, W>,
    ) -> Result<(), anyhow::Error> {
        self.commitment_scheme.read_commitment(channel)
    }
```